### PR TITLE
Make `mypy` checks opt-out instead of opt-in, fix several type issues

### PR DESCRIPTION
**Pull request by lostella**
*Description of changes:*
- Fix several type issues through the package
- Mark many other failing checks as "ignore", either because there's too much code to go through otherwise (rotbaum), or because it would be impractical to ensure checks passing
- Update setup.py to use an opt-out mechanism for type checking, instead of opt-in
- Remove all `.typesafe` files, add a `.typeunsafe` file to `gluonts.nursery` to opt it out
- Marking as BREAKING because some code moved from `gluonts.torch.distributions` to `gluonts.torch.model.mqf2`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

**Comment by lostella** [comment]
Todos:
- [x] get rid of a circular import I introduced (may need to break some import path there, but for good reasons)
- [x] solve mypy complaints about "duplicate modules" (see following, also [here](https://github.com/awslabs/gluonts/actions/runs/6627134500/job/18001440630?pr=3027#step:6:68))
```
src/gluonts/torch/model/deep_npts/__init__.py: error: Duplicate module named "gluonts.torch.model.deep_npts" (also at "/home/runner/work/gluonts/gluonts/src/gluonts/torch/model/deep_npts/__init__.py")
src/gluonts/torch/model/deep_npts/__init__.py: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#mapping-file-paths-to-modules for more info
src/gluonts/torch/model/deep_npts/__init__.py: note: Common resolutions include: a) using `--exclude` to avoid checking one of them, b) adding `__init__.py` somewhere, c) using `--explicit-package-bases` or adjusting MYPYPATH
```
- [x] make the type-checking mechanism opt-out, and opt-out problematic packages in `gluonts.nursery`

File: src/gluonts/model/forecast.py [hunk 1]
```diff
@@ -247,9 +247,12 @@ class Forecast:
     item_id: Optional[str]
     info: Optional[Dict]
     prediction_length: int
-    mean: np.ndarray
     _index = None
 
+    @property
+    def mean(self) -> np.ndarray:
+        raise NotImplementedError()
```
Comment (shchur) [inline src/gluonts/model/forecast.py:254]: Nit:
```suggestion
        raise NotImplementedError
```

File: src/gluonts/torch/model/lightning_util.py [hunk 2]
```diff
@@ -18,5 +18,6 @@
 
 def has_validation_loop(trainer: pl.Trainer):
     if version.parse(pl.__version__) < version.parse("2.0.0"):
+        assert hasattr(trainer._data_connector, "_val_dataloader_source")
```
Comment (shchur) [inline src/gluonts/torch/model/lightning_util.py]: We have changed the dependency range for lightning to `>=2.0`, can we just drop this check completely?

File: src/gluonts/torch/model/wavenet/estimator.py [hunk 3]
```diff
@@ -96,7 +96,7 @@ def __init__(
         num_batches_per_epoch: int = 50,
         num_parallel_samples: int = 100,
         negative_data: bool = False,
-        trainer_kwargs: Dict[str, Any] = None,
+        trainer_kwargs: Dict[str, Any] = dict(),
```
Comment (shchur) [inline src/gluonts/torch/model/wavenet/estimator.py]: I think we shouldn't use mutable types as default arguments. Maybe change this to 
```python
        trainer_kwargs: Optional[Dict[str, Any]] = None,
```
and also undo the change on L181 below?

File: src/gluonts/torch/model/wavenet/module.py [hunk 4]
```diff
@@ -329,24 +326,20 @@ def base_net(
...
-                queues is not None
-            ), "Queues cannot be empty in prediction mode!"
-
         skip_outs = []
         queues_next = []
         out = inputs
         for i, layer in enumerate(self.residuals):
             skip, out = layer(out)
-            if prediction_mode:
+            if queues is not None:
```
Comment (shchur) [inline src/gluonts/torch/model/wavenet/module.py:342]: That's a nice simplification!

File: src/gluonts/torch/model/wavenet/estimator.py (hunk 3 above)
Comment (lostella) [reply src/gluonts/torch/model/wavenet/estimator.py]: done

File: src/gluonts/torch/model/wavenet/module.py (hunk 4 above)
Comment (lostella) [reply src/gluonts/torch/model/wavenet/module.py:342]: and the fact is: type checking surfaced this, we should have it opt-out instead of opt-in

File: src/gluonts/torch/model/lightning_util.py (hunk 2 above)
Comment (lostella) [reply src/gluonts/torch/model/lightning_util.py]: done

File: src/gluonts/model/forecast.py (hunk 1 above)
Comment (lostella) [reply src/gluonts/model/forecast.py:254]: Keeping it this way for consistency with the code below: if linters/style checkers are fine with it, I'm fine with it!

**Review by jaheba — DISMISSED** [review]

**Review by jaheba — APPROVED** [review]
